python/cudf/cudf/tests/test_indexing.py

```diff
@@ -668,6 +668,20 @@ def test_dataframe_iloc(nelem):
     assert_eq(gdf.iloc[np.array([0])], pdf.loc[np.array([0])])
 
 
+def test_dataframe_iloc_single_row_downcast():
+    # checks to make sure a single row is downcast to a series
+    # even if the resulting dtype has to be string
+    gdf = cudf.DataFrame({"a": [1, 2, 3], "b": ["a", "b", "c"]})
+    pdf = gdf.to_pandas()
+
+    got = gdf.iloc[0]
+    expect = pdf.iloc[0]
+    # at this point expect is object dtype and contains an int and a string
+    expect = expect.map(str)
```

Comment:
Would the eventual goal be to preserve the `int` and `str` in the resulting `Series` from `gdf.iloc[0]`? I can imagine it being surprising having the `1` being cast to `"1"`.

Reply:
Yeah, this certainly is a tradeoff. The main problem stems from the fact that pandas can create a series with a true object dtype, where cuDF can not. This makes perfect sense for taking a row of a dataframe with mixed dtypes as the individual elements of the resulting series retain the dtypes of the columns they came from. 

```python
df = pd.DataFrame({
    'a': [1, 2, 3],
    'b': ['a', 'b', 'c']
})
df.iloc[0]
a    1
b    a
Name: 0, dtype: object
```
cuDF series objects don't support object though so we end up being forced to cast to the common type of the columns in the source dataframe. But it seems the machinery that did that decided that if non numeric types were present, to instead return a dataframe object with one row to preserve dtypes rather than cast. 

This creates a really weird API situation where the same workflow can either work or error depending on the exact type of data present. The user that brought this to my attention was trying to obtain a scalar from a dataframe by doing 
```python
df.iloc[0]['col_name']
```
which will return either a scalar or a `Series` depending on the dtypes in `df`. There's two ways out of this in my mind:

1. We can always return a dataframe object when a single row is requested. This makes cuDF consistent regardless of the data, but makes the pandas and cuDF APIs diverge since pandas will return a series. This isn't perfect but perhaps we could note this loudly in the docs.
2. We can extend the casting rule to all scalar types and return a series. Admittedly this has the drawbacks you point out which, now that I think about it, are even worse when we think beyond object type to datetime, timedelta, or structured types.

This change originally proposed (2) however thinking through it now I almost think bailing to (1) is the lesser of two evils. Open to thoughts here.

Reply:
Ah thanks for the context; I didn't know that cuDF didn't support an arbitrary `object` type. 

I like @shwina suggestion on Slack on raising an exception (`NotImplementedError`?) if the user is trying to do this and there isn't the same dtype across the row.

Reply:
I agree with @mroeschke and @shwina, either approach will be incompatible with pandas. So I would go for `NotImplementedError`. In fact, I am -epsilon on the existing implementation that tries to unify all numeric columns into a series:
```python
df = cudf.DataFrame({"a": np.asarray([1], dtype=float), "b": np.asarray([2**63 - 1], dtype=int)})
assert df.iloc[0][1].astype(int) == df.iloc[0, 1] => AssertionError
```